features/dd-sdk-android-session-replay-compose/src/main/kotlin/com/datadog/android/sessionreplay/compose/internal/mappers/semantics/ImageSemanticsNodeMapper.kt

```diff
@@ -0,0 +1,140 @@
+        }
+        val imageWireframe = applicationContext?.let { context ->
+            if (bitmapInfo != null) {
+                parentContext.imageWireframeHelper.createImageWireframe(
+                    id = semanticsNode.id.toLong(),
+                    globalBounds = bounds,
+                    applicationContext = context,
+                    bitmap = bitmapInfo.bitmap,
+                    density = parentContext.density,
+                    isContextualImage = bitmapInfo.isContextualImage,
+                    imagePrivacy = ImagePrivacy.MASK_NONE,
+                    asyncJobStatusCallback = asyncJobStatusCallback,
+                    clipping = null,
+                    shapeStyle = null,
+                    border = null
+                )
+            } else {
+                null
+            }
+        }
+        return imageWireframe?.let {
+            ComposeWireframe(
+                imageWireframe,
+                null
+            )
+        }
+    }
+
+    private fun resolveSemanticsPainter(
+        semanticsNode: SemanticsNode
+    ): BitmapInfo? {
+        var isContextualImage = false
+        var painter = tryParseLocalImagePainter(semanticsNode)
+        if (painter == null) {
+            painter = tryParseAsyncImagePainter(semanticsNode)
+            if (painter != null) {
+                isContextualImage = true
+            }
+        }
+        // TODO RUM-6535: support more painters.
+        val bitmap = when (painter) {
+            is BitmapPainter -> tryParseBitmapPainterToBitmap(painter)
+            is VectorPainter -> tryParseVectorPainterToBitmap(painter)
+            else -> {
+                null
+            }
+        }
+
+        val newBitmap = bitmap?.let {
+            @Suppress("UnsafeThirdPartyFunctionCall") // isMutable is always false
+            it.copy(it.config, false)
+        }
+        return newBitmap?.let {
+            BitmapInfo(it, isContextualImage)
+        }
+    }
+
+    private fun resolveApplicationContext(semanticsNode: SemanticsNode): Context? {
+        val owner = LayoutNodeOwnerField?.getSafe(semanticsNode.layoutInfo) as? ViewGroup
+        return owner?.context
+    }
+
+    private fun tryParseVectorPainterToBitmap(vectorPainter: VectorPainter): Bitmap? {
+        val vector = ComposeReflection.VectorField?.getSafe(vectorPainter)
+        val cacheDrawScope = ComposeReflection.CacheDrawScopeField?.getSafe(vector)
+        val mCachedImage = ComposeReflection.CachedImageField?.getSafe(cacheDrawScope)
+        return BitmapField?.getSafe(mCachedImage) as? Bitmap
+    }
+
+    private fun tryParseBitmapPainterToBitmap(bitmapPainter: BitmapPainter): Bitmap? {
+        val image = ImageField?.getSafe(bitmapPainter)
+        return BitmapField?.getSafe(image) as? Bitmap
+    }
+
+    private fun tryParseLocalImagePainter(semanticsNode: SemanticsNode): Painter? {
+        val modifier = semanticsNode.layoutInfo.getModifierInfo().firstOrNull {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❔ question:** ‏ Are these all reflection calls ? Aren't we concerned that performing all these reflection calls every time we take a snapshot is going to impact app perfs ?

Reply:
this one is not, `semanticsNode`, `layoutInfo` and `getModifierInfo` are all public apis that we have access. But indeed we have some reflections in compose SR, we will setup some benchmarking to measure the performance later.